pkg/iam_verify/iam_verify.go

```diff
@@ -0,0 +1,361 @@
+	"ec2:AuthorizeSecurityGroupIngress",
+	"ec2:CreateInternetGateway",
+	"ec2:CreateNatGateway",
+	"ec2:CreateRoute",
+	"ec2:CreateRouteTable",
+	"ec2:CreateSecurityGroup",
+	"ec2:CreateSubnet",
+	"ec2:CreateTags",
+	"ec2:CreateVpc",
+	"ec2:DeleteInternetGateway",
+	"ec2:DeleteNatGateway",
+	"ec2:DeleteNetworkInterface",
+	"ec2:DeleteRoute",
+	"ec2:DeleteRouteTable",
+	"ec2:DeleteSecurityGroup",
+	"ec2:DeleteSubnet",
+	"ec2:DeleteVpc",
+	"ec2:DescribeAccountAttributes",
+	"ec2:DescribeAddresses",
+	"ec2:DescribeAvailabilityZones",
+	"ec2:DescribeImages",
+	"ec2:DescribeInstanceAttribute",
+	"ec2:DescribeInstances",
+	"ec2:DescribeInternetGateways",
+	"ec2:DescribeNatGateways",
+	"ec2:DescribeNetworkAcls",
+	"ec2:DescribeNetworkInterfaces",
+	"ec2:DescribeRegions",
+	"ec2:DescribeRouteTables",
+	"ec2:DescribeSecurityGroups",
+	"ec2:DescribeSubnets",
+	"ec2:DescribeTags",
+	"ec2:DescribeVolumes",
+	"ec2:DescribeVpcAttribute",
+	"ec2:DescribeVpcClassicLink",
+	"ec2:DescribeVpcClassicLinkDnsSupport",
+	"ec2:DescribeVpcs",
+	"ec2:DetachInternetGateway",
+	"ec2:DisassociateRouteTable",
+	"ec2:ModifyNetworkInterfaceAttribute",
+	"ec2:ModifyVpcAttribute",
+	"ec2:ReleaseAddress",
+	"ec2:ReplaceRouteTableAssociation",
+	"ec2:RevokeSecurityGroupEgress",
+	"ec2:RevokeSecurityGroupIngress",
+	"ec2:RunInstances",
+	"ec2:TerminateInstances",
+	"elasticloadbalancing:AddTags",
+	"elasticloadbalancing:CreateLoadBalancer",
+	"elasticloadbalancing:DeleteLoadBalancer",
+	"elasticloadbalancing:DescribeLoadBalancerAttributes",
+	"elasticloadbalancing:DescribeLoadBalancers",
+	"elasticloadbalancing:DescribeTags",
+	"elasticloadbalancing:ModifyLoadBalancerAttributes",
+	"iam:AddRoleToInstanceProfile",
+	"iam:CreateInstanceProfile",
+	"iam:CreateRole",
+	"iam:DeleteInstanceProfile",
+	"iam:DeleteRole",
+	"iam:DeleteRolePolicy",
+	"iam:GetInstanceProfile",
+	"iam:GetRole",
+	"iam:ListInstanceProfiles",
+	"iam:ListRolePolicies",
+	"iam:PassRole",
+	"iam:PutRolePolicy",
+	"iam:RemoveRoleFromInstanceProfile",
+	"route53:ChangeResourceRecordSets",
+	"route53:ChangeTagsForResource",
+	"route53:CreateHostedZone",
+	"route53:DeleteHostedZone",
+	"route53:GetChange",
+	"route53:GetHostedZone",
+	"route53:ListHostedZones",
+	"route53:ListResourceRecordSets",
+	"route53:ListTagsForResource",
+	"route53:UpdateHostedZoneComment",
+	"s3:CreateBucket",
+	"s3:DeleteBucket",
+	"s3:DeleteObject",
+	"s3:GetAccelerateConfiguration",
+	"s3:GetBucketCors",
+	"s3:GetBucketEncryption",
+	"s3:GetBucketLifecycle",
+	"s3:GetBucketLocation",
+	"s3:GetBucketLogging",
+	"s3:GetBucketRequestPayment",
+	"s3:GetBucketTagging",
+	"s3:GetBucketVersioning",
+	"s3:GetBucketWebsite",
+	"s3:GetEncryptionConfiguration",
+	"s3:GetLifecycleConfiguration",
+	"s3:GetObject",
+	"s3:GetObjectTagging",
+	"s3:GetReplicationConfiguration",
+	"s3:HeadBucket",
+	"s3:ListAllMyBuckets",
+	"s3:ListBucket",
+	"s3:PutBucketAcl",
+	"s3:PutBucketTagging",
+	"s3:PutObject",
+	"s3:PutObjectAcl",
+	"s3:PutObjectTagging",
```

Comment:
I collected it by hand, yes. I created a new IAM user and added these permissions based on the errors shown in CloudTrail while running the installer. It took a lot of trial/error and digging through logs. It would be great if we had a way to automate this.

We do need a way to maintain this list too, that's a very good point. Offhand I'm not sure how. Maybe something in CI? A test that creates an IAM user with the list of permissions given, and attempts to run the installer. Then, when someone adds code that requires a new permission, they'll see the error message and update the list.

If the operators declared the minimum required permissions to run, that would be useful too. I imagine we'll be responsible for telling our users which permissions they'll need in their cloud provisioning account. Or at the very least, giving them some meaningful error feedback when provisioning fails due to account permissions.

Reply:
Can you create a follow up ticket for what @wking pointed out regarding other operators declaring their permissions.

Reply:
@spangenberg @wking if I'm reading this correctly this code verifies perms for the user to run the installer. The perms needed by machine-api-operator are defined in the iam attached to the instance, i.e https://github.com/openshift/installer/blob/master/data/data/aws/master/main.tf#L48

Reply:
> The perms needed by machine-api-operator along with any other operator are defined in the iam attached to the instance...

We're hoping to move away from node-scoped permissions soon and switch to pod-scoped (or container-scoped?) permissions.  Once that happens, we'll need to know which permissions each pod/container is going to need, so the permission allocator doesn't hand out more (or less) than required.

Reply:
This jira ticket should take care of maintaining the list https://jira.coreos.com/browse/CORS-800?focusedCommentId=53703&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-53703

Let me know if there's any additional action I need to take. Thanks.